File: scripts/hero-contrast.mjs
import { chromium, devices } from "playwright";
import { writeFileSync } from "node:fs";

/*
 * Measures the contrast of the hero text against the photograph actually
 * behind it. The text is hidden, the band under each line is screenshotted,
 * and white is compared against the brightest 10% of those pixels — the
 * patches of sky and leaf that make a line hard to read.
 */
const ORIGIN = process.argv[2] ?? "http://localhost:3100";
const b = await chromium.launch({ channel: "chrome" });
const out = {};
for (const [label, cfg] of [
  ["desktop", { viewport: { width: 1440, height: 900 } }],
  ["phone", devices["iPhone 13"]],
]) {
  const c = await b.newContext(cfg);
  const p = await c.newPage();
  await p.goto(`${ORIGIN}/`, { waitUntil: "domcontentloaded" });
  await p.evaluate(() => localStorage.setItem("swarma.consent.v1", "accepted"));
  await p.goto(`${ORIGIN}/`, { waitUntil: "networkidle" });
  await p.waitForTimeout(3800);
  const boxes = await p.evaluate(() => {
    const hero = document.querySelector('[data-hero="dark"]');
    const pick = { kicker: hero.querySelector(".kicker"), h1: hero.querySelector("h1"), lede: hero.querySelector("p.max-w-md") };
    const r = {};
    for (const [k, el] of Object.entries(pick)) {
      const b = el.getBoundingClientRect();
      r[k] = { x: b.x, y: b.y, w: b.width, h: b.height };
      el.style.visibility = "hidden";
    }
    // header too — it sits on the same photograph
    const nav = document.querySelector("header nav") ?? document.querySelector("header");
    const hb = nav.getBoundingClientRect();
    r.nav = { x: hb.x, y: hb.y, w: hb.width, h: hb.height };
    document.querySelector("header").style.visibility = "hidden";
    return r;
  });
  await p.waitForTimeout(300);
  const res = {};
  for (const [k, bx] of Object.entries(boxes)) {
    const buf = await p.screenshot({ clip: { x: bx.x, y: bx.y, width: Math.max(1, bx.w), height: Math.max(1, bx.h) } });
    res[k] = buf.toString("base64");
  }
  out[label] = res;
  await c.close();
}
await b.close();
writeFileSync("scratch/hero-bands.json", JSON.stringify(out));
console.log("captured");

File: scripts/hero-contrast.py
import json, base64, io
from PIL import Image
def lin(v):
    v /= 255; return v / 12.92 if v <= 0.04045 else ((v + 0.055) / 1.055) ** 2.4
def lum(p): return 0.2126 * lin(p[0]) + 0.7152 * lin(p[1]) + 0.0722 * lin(p[2])
need = {"kicker": 4.5, "lede": 4.5, "nav": 4.5, "h1": 3.0}
data = json.load(open("scratch/hero-bands.json"))
worst = 99
for device, bands in data.items():
    for k, b64 in bands.items():
        im = Image.open(io.BytesIO(base64.b64decode(b64))).convert("RGB")
        ls = sorted(lum(p) for p in im.getdata())
        p90 = ls[int(len(ls) * 0.9)]
        ratio = 1.05 / (p90 + 0.05)
        ok = ratio >= need[k]
        worst = min(worst, ratio / need[k])
        print(f"{device:8} {k:7} white vs brightest 10%: {ratio:5.2f}:1  need {need[k]}  {'ok' if ok else 'LOW'}")
print("worst margin:", round(worst, 2))
